pandas/core/algorithms.py

```diff
@@ -988,7 +988,9 @@ def _get_data_algo(values, func_map):
     elif is_unsigned_integer_dtype(values):
         f = func_map['uint64']
         values = _ensure_uint64(values)
-
+    elif is_categorical(values) and values._ordered:
```

Comment:
that else should be ``values = np.array(self)``, which *will* order by the actual objects whether integers, objects or whatever. Its NOT correct to ``astype(object``, rather to use the natural object ordering which is dependent on the dtype (but that is already handled correctly). The key point is that if we are ordered we return a float array based on the codes, else return self.